packages/grafana-data/src/field/fieldOverrides.ts

```diff
@@ -0,0 +1,213 @@
+  replaceVariables: InterpolateFunction,
+  theme: GrafanaTheme,
+  isUtc?: boolean
+): DataFrame[] {
+  if (!source) {
+    return data;
+  }
+  let range: GlobalMinMax | undefined = undefined;
+
+  // Prepare the Matchers
+  const override: OverrideProps[] = [];
+  if (source.overrides) {
+    for (const rule of source.overrides) {
+      const info = fieldMatchers.get(rule.matcher.id);
+      if (info) {
+        override.push({
+          match: info.get(rule.matcher),
+          properties: rule.properties,
+        });
+      }
+    }
+  }
+
+  return data.map((frame, index) => {
+    let name = frame.name;
+    if (!name) {
+      name = `Series[${index}]`;
+    }
+
+    const fields = frame.fields.map(field => {
+      // Config is mutable within this scope
+      const config: FieldConfig = { ...field.config } || {};
+      if (field.type === FieldType.number) {
+        setFieldConfigDefaults(config, source.defaults);
+      }
+
+      // Find any matching rules and then override
+      for (const rule of override) {
+        if (rule.match(field)) {
+          for (const prop of rule.properties) {
+            setDynamicConfigValue(config, {
+              value: prop,
+              config,
+              field,
+              data: frame,
+              replaceVariables,
+            });
+          }
+        }
+      }
+
+      // Set the Min/Max value automatically
+      if (field.type === FieldType.number) {
+        if (!isNumber(config.min) || !isNumber(config.max)) {
+          if (!range) {
+            range = findNumericFieldMinMax(data);
+          }
+          if (!isNumber(config.min)) {
+            config.min = range.min;
+          }
+          if (!isNumber(config.max)) {
+            config.max = range.max;
+          }
+        }
+      }
+
+      return {
+        ...field,
+
+        // Overwrite the configs
+        config,
+
+        // Set the display processor
+        processor: getDisplayProcessor({
+          type: field.type,
+          config: config,
+          theme,
+          isUtc,
+        }),
+      };
+    });
+
+    return {
+      ...frame,
+      fields,
+      name,
+    };
+  });
+}
+
+interface DynamicConfigValueOptions {
+  value: DynamicConfigValue;
+  config: FieldConfig;
+  field: Field;
+  data: DataFrame;
+  replaceVariables: InterpolateFunction;
+}
+
+const numericFieldProps: any = {
+  decimals: true,
+  min: true,
+  max: true,
+};
+
+function prepareConfigValue(key: string, input: any, options?: DynamicConfigValueOptions): any {
+  if (options) {
+    // TODO template variables etc
+  }
+
+  if (numericFieldProps[key]) {
+    const num = toNumber(input);
+    if (isNaN(num)) {
+      return null;
+    }
+    return num;
+  } else if (input) {
+    // skips empty string
+    if (key === 'unit' && input === 'none') {
+      return null;
+    }
+  }
+  return input;
+}
+
+export function setDynamicConfigValue(config: FieldConfig, options: DynamicConfigValueOptions) {
```

Comment:
Not sure we need config as an argument here as it's passed via `options` too. And mutating the config feelt unintuitive to me at first sight. But this is just a nit.

Reply:
ya -- I went back and forth on mutating vs always setting one property.  I can go back, but let me explain what I was thinking:

In the beginning of the `applyFieldOverrides` loop, the config is copied and then all the rules are attached to it -- inside a few more loops.

The setXXX functions are really internal to this file -- they are only exported so the tests can see them.  Note they are not exported in the `index.ts` file, only:

```
export { applyFieldOverrides } from './fieldOverrides';
```

Reply:
@ryantxu yup, got it when reading the `applyFieldOverrides` fn:)